### PR TITLE
Add Service layer and api endpoints for Interventions

### PR DESCRIPTION
**Pull request by karenmillermoj**
## What does this pull request do?

Adds the ability to find interventions.  Also adds basic pact test fixes

## What is the intent behind these changes?

Build on the intervention model to provide access to intervention data

### CODE REVIEW

**Comment by karenmillermoj** [comment]
Note. Additional tests are still to be added for this.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/InterventionService.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+class InterventionService(val repository: InterventionRepository) {
+
+  fun getIntervention(id: UUID): Intervention? {
+    return repository.findByIdOrNull(id)
+  }
+  fun getInterventionsForServiceProvider(id: String): List<Intervention> {
+    return repository.findByDynamicFrameworkContractServiceProviderId(id)
+  }
+  fun createIntervention(intervention: Intervention): Intervention {
+    return repository.save(intervention)
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/InterventionService.kt:19]: there are quite a few entities that need to exist for a full intervention to be stored:

- ServiceCategory
- ServiceProvider
- ContractEligibility
- NPSRegion / PCCRegion
- DynamicFrameworkContract

i think we should be ensuring that all these things are persisted here. which follows the example of `ReferralService.sendDraftReferral` and `ReferralService.createDraftReferral`.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.dto.InterventionDTO
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service.InterventionService
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.utils.StringToUUIDConverter
+
+@RestController
+class InterventionController(
+  private val interventionService: InterventionService,
+) {
+
+  @GetMapping("/intervention/{id}")
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt:17]: @sldblog do we want the endpoints to differentiate between structured and DF interventions?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/InterventionService.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.InterventionRepository
+import java.util.UUID
+
+@Service
+class InterventionService(val repository: InterventionRepository) {
+
+  fun getIntervention(id: UUID): Intervention? {
+    return repository.findByIdOrNull(id)
+  }
+  fun getInterventionsForServiceProvider(id: String): List<Intervention> {
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/InterventionService.kt:15]: what do you think about type aliasing the service provider id?

File: src/main/resources/db/local/V99_0__draft_referrals.sql [hunk 4]
```diff
@@ -6,7 +6,14 @@ insert into referral (id, created_at, service_usercrn, completion_deadline, crea
...
        ('dfb64747-f658-40e0-a827-87b4b0bdcfed', '2020-12-07 20:45:21.986389+00', 'X456234', '2021-03-01', '8751622134', null),
        ('d496e4a7-7cc1-44ea-ba67-c295084f1962', '2020-12-24 09:32:32.871623+00', 'X542343', '2021-01-30', '2500128586', '428ee70f-3001-4399-95a6-ad25eaaede16'),
-       ('1219a064-709b-4b6c-a11e-10b8cb3966f6', '2021-01-12 14:46:21.987234+00', 'X862134', null, '2500128586', '428ee70f-3001-4399-95a6-ad25eaaede16');
+       ('1219a064-709b-4b6c-a11e-10b8cb3966f6', '2021-01-12 14:46:21.987234+00', 'X862134', null, '2500128586', '428ee70f-3001-4399-95a6-ad25eaaede16'),
+       ('037cc90b-beaa-4a32-9ab7-7f79136e1d27', '2021-01-12 14:46:21.987234+00', 'X872391', null, '2500128586', '428ee70f-3001-4399-95a6-ad25eaaede16');
 
 insert into referral_service_user_data (referral_id, disabilities, dob, ethnicity, title, first_name, last_name, preferred_language, religion_or_belief, gender)
 values ('1219a064-709b-4b6c-a11e-10b8cb3966f6', '{Autism spectrum condition}', TO_DATE('1980-08-10', 'YYYY-MM-DD'), 'White British', 'Mr', 'Alex', 'River', 'English', 'None', 'Male');
+
+insert into referral_desired_outcome (referral_id, desired_outcome_id)
```
Comment (tomsmyers) [inline src/main/resources/db/local/V99_0__draft_referrals.sql:15]: are these related to this PR?

Comment (karenmillermoj) [reply src/main/resources/db/local/V99_0__draft_referrals.sql:15]: They are required for some of the Pact Tests.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/InterventionService.kt (hunk 1 above)
Comment (karenmillermoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/InterventionService.kt:19]: Totally agree - as part of the CSV work (separate Pull Request) they will exist in there. It has a separate DTO too - as it requires more information than the one the UI requires. ( I am happy to remove this method for now unless you're ok with the fact that the next PR will pick this up.)

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/InterventionHelper.kt [hunk 5]
```diff
@@ -0,0 +1,93 @@
+package uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity
+
+import java.time.LocalDate
+import java.time.OffsetDateTime
+import java.util.UUID
+
+class InterventionHelper {
```
Comment (sldblog) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/InterventionHelper.kt:7]: Since all of the functions here are samples, what do you think about calling this class `InterventionFixtures` or `Fixtures`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt (hunk 2 above)
Comment (sldblog) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt:17]: Given we're expected to support discovering all interventions and services, I think we're currently ok with a common prefix.

These are implied to be published or available interventions (vs in-progress, unlisted, under QA) -- what do you think about using `/published-interventions/{id}`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  fun getInterventionByID(@PathVariable id: String): InterventionDTO {
+    val uuid = StringToUUIDConverter.parseID(id)
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt:19]: Would it be possible to hook that conversion into the request mapper so we only have to do `getInterventionByID(@PathVariable id: UUID)`? It looks like that'd save a _lot_ of repetition in the controllers.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+    return interventionService.getIntervention(uuid)
+      ?.let { InterventionDTO.from(it) }
+      ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "intervention not found [id=$uuid]")
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt:23]: We have the opportunity to drop the message if you'd like

- "intervention not found" is implicit from 404 Not Found on a call to /intervention/{id}
- the id is also implicit in the caller context

**Review by sldblog — COMMENTED** [review]
Looks good, just a couple of question and a reply 👍

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt (hunk 7 above)
Comment (karenmillermoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt:23]: true  - will update

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt (hunk 6 above)
Comment (karenmillermoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt:19]: Yes - updated

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt (hunk 2 above)
Comment (karenmillermoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt:17]: hmm...good point - maybe we should discuss this to decide

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/InterventionHelper.kt (hunk 5 above)
Comment (karenmillermoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/InterventionHelper.kt:7]: Sure.  It looks like the Referral tests could make sure of this class so a more general name looks preferable. How about just "SampleData" ?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt (hunk 6 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt]: this pattern is reused a lot. would be good to change all the other occurances (outside this PR).

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt (hunk 7 above)
Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt]: this follows existing patterns. the message is formatted into an error response that can be used by the UI. would prefer to keep it consistent.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/InterventionService.kt (hunk 3 above)
Comment (karenmillermoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/InterventionService.kt:15]: yep - done

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt (hunk 6 above)
Comment (karenmillermoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt]: definitely

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt (hunk 7 above)
Comment (karenmillermoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/InterventionController.kt]: ok - will put back in

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/ServiceProvider.kt [hunk 8]
```diff
@@ -7,7 +7,8 @@ import javax.validation.constraints.NotNull
 @Entity
 data class ServiceProvider(
   // Service Provider id maps to the hmpps-auth field Group#groupCode
-  @NotNull @Id val id: String,
+  @NotNull @Id val id: AuthGroupID,
   @NotNull val name: String,
   @NotNull val incomingReferralDistributionEmail: String,
 )
+typealias AuthGroupID = String
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/entity/ServiceProvider.kt:14]: this could use a comment explaining what it is

**Review by tomsmyers — APPROVED** [review]
🤘🏼
